### DISCUSSION TITLE
How to install Pybricks on 51515 hub?

### DISCUSSION THREAD

**Discussion by BorisBrock** [pybricks/support#763] [Q&A]
All the instructions here regarding the 51515 hub/Spike Prime seem to have vanished.
Can Pybricks be installed on those hubs? If so, how?

Bonus question: is sound output working? (the original Lego firmware had plenty of nice robot sounds to use...)

**Comment by laurensvalk** [comment]
> Can Pybricks be installed on those hubs? If so, how?

Yes! Since last week you can try it via https://beta.pybricks.com/

And you're right, we still need to update our website to reflect this. :+1:

**Reply by BorisBrock** [reply]
I just tried to install the firmware with the beta version:

- Copied the udev .rules file to /etc/udev/rules.d/99-pybricksdev.rules (as described here https://github.com/pybricks/support/discussions/688)
- Restarted laptop
- Connected the hub and put it into boot loader mode

Sadly the Pybricks website still cannot flash the hub. All I get is "No compatible devices found".
Is there any way I can debug this to find out what is going wrong?

PS: I'm using Manjaro Linux

**Reply by laurensvalk** [reply]
With the hub in DFU mode, what do you see with `lsusb`?

**Reply by BorisBrock** [reply]
It seems to be there:

> Bus 003 Device 012: ID 0694:0011 Lego Group LEGO Technic Large Hub in DFU Mode

**Reply by laurensvalk** [reply]
That looks good. Does it help if you:

```
sudo adduser $USER dialout
```

And reboot?

**Reply by laurensvalk** [reply]
If that doesn't help, try adding a file ``77-dfu.rules`` with:

```
# SPIKE PRIME PYBRICKS USB DFU
ATTRS{idProduct}=="0008", ATTRS{idVendor}=="0694", MODE="666", GROUP="dialout"
ATTRS{idProduct}=="000c", ATTRS{idVendor}=="0694", MODE="666", GROUP="dialout"
ATTRS{idProduct}=="0011", ATTRS{idVendor}=="0694", MODE="666", GROUP="dialout"

# 0694:0009 - SPIKE PRIME PYBRICKS USB
ATTRS{idVendor}=="0694", ENV{ID_MM_DEVICE_IGNORE}="1"
ATTRS{idVendor}=="0694", ENV{ID_MM_PORT_IGNORE}="1"
ATTRS{idVendor}=="0694", ENV{MTP_NO_PROBE}="1"
SUBSYSTEMS=="usb", ATTRS{idVendor}=="0694", MODE:="0666"
KERNEL=="ttyACM*", ATTRS{idVendor}=="0694", MODE:="0666"
```

The `99-pybricksdev.rules` did not work for me previously on Ubuntu 18.04 but the ones above did. Maybe they help you too.

Although today on `Ubuntu 22.04` the `99-pybricksdev.rules` seem to work fine for me.

**Reply by BorisBrock** [reply]
Added myself to the dialout group and did a reboot. -> Its working now :-)
Thanks a lot!

The firmware flashing was successful. Now I can't get the hub to connect via bluetooth, but that's probably a different story... ;-)

**Reply by laurensvalk** [reply]
Thanks for confirming! I've added a note about `dialout` to the installation reference.

Feel free to open a new issue about any bluetooth issues :)

**Reply by dlech** [reply]
The `dialout` group only makes a difference for the USB serial port on the official LEGO firmware using existing udev rules for generic USB serial ports. It won't have any effect on the DFU bootloader connection. (and MODE of 666 gives the user the same permission as the group so the user doesn't have to be in a specific group).

The thing that probably made the difference here is the `MODE` option vs. the `TAG` option. The latter requires systemd support.

**Reply by laurensvalk** [reply]
Apparently dialout fixed it for @BorisBrock , not the alternate rules, right?

I had to do it as well on my fresh Ubuntu 22.04 install to make DFU work, even with the 99 udev rules.

**Reply by dlech** [reply]
How strange. I wonder if Chromium is looking for the dialout group or something like that?

**Reply by laurensvalk** [reply]
Quite possibly. I seem to remember pybricksdev worked right away, but I'm not sure.

**Reply by dlech** [reply]
Out of curiosity, I removed myself from the `dialout` group on Ubuntu 22.04 and rebooted. DFU still works - tested with both pybricksdev and Pybricks Code. After trying `pybricksdev dfu backup`, the hub didn't show up in Pybricks Code until I rebooted the hub (SPIKE Prime).

**Comment by dlech** [comment]
> is sound output working?

We have beeps, but can't play .wav files yet.

https://docs.pybricks.com/en/latest/hubs/primehub.html#pybricks.hubs.PrimeHub.speaker.volume
